File: 01-Programming-Languages/C++/Data-Structures/Graph/graph_adjacency_list.cpp
/*
Problem: Graph as Adjacency List
Implement a graph using adjacency list and perform DFS traversal.

Approach:
- Array of linked lists to represent edges

Time Complexity: O(V + E)
Space Complexity: O(V + E)

Example:
Input: 4 vertices, edges: (0,1),(0,2),(1,2),(2,3)
Output: DFS: 0 1 2 3
*/

#include <iostream>
#include <vector>
#include <list>
using namespace std;

class Graph
{
    int V;
    vector<list<int>> adj;

    void dfs_util(int v, vector<bool>& visited)
    {
        visited[v] = true;
        cout << v << " ";
        for (int u : adj[v])
            if (!visited[u]) dfs_util(u, visited);
    }

public:
    Graph(int vertices) : V(vertices), adj(vertices) {}

    void add_edge(int src, int dest)
    {
        adj[src].push_back(dest);
        adj[dest].push_back(src);
    }

    void dfs(int start)
    {
        vector<bool> visited(V, false);
        dfs_util(start, visited);
        cout << endl;
    }
};

int main()
{
    Graph g(4);
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 2);
    g.add_edge(2, 3);
    cout << "DFS from 0: ";
    g.dfs(0);
    return 0;
}

File: 01-Programming-Languages/JavaScript/Data-Structures/Graph/graphAdjacencyList.js
/*
Problem: Graph Adjacency List
Description: Implement a graph using an adjacency list with addVertex, addEdge,
removeEdge, removeVertex, and display methods.

Approach:
- Use a Map of vertices to arrays of neighbors

Time Complexity: O(1) add, O(V+E) display
Space Complexity: O(V+E)

Example:
Input: addVertex(0), addVertex(1), addEdge(0,1), display()
Output: 0 -> [1], 1 -> [0]
*/

class Graph {
  constructor() {
    this.adjList = new Map();
  }

  addVertex(vertex) {
    if (!this.adjList.has(vertex)) {
      this.adjList.set(vertex, []);
    }
  }

  addEdge(v1, v2) {
    this.adjList.get(v1).push(v2);
    this.adjList.get(v2).push(v1);
  }

  removeEdge(v1, v2) {
    this.adjList.set(v1, this.adjList.get(v1).filter(v => v !== v2));
    this.adjList.set(v2, this.adjList.get(v2).filter(v => v !== v1));
  }

  removeVertex(vertex) {
    for (const neighbor of this.adjList.get(vertex)) {
      this.removeEdge(vertex, neighbor);
    }
    this.adjList.delete(vertex);
  }

  display() {
    for (const [vertex, neighbors] of this.adjList) {
      console.log(`${vertex} -> [${neighbors.join(', ')}]`);
    }
  }
}

const graph = new Graph();
graph.addVertex(0);
graph.addVertex(1);
graph.addVertex(2);
graph.addEdge(0, 1);
graph.addEdge(0, 2);
graph.addEdge(1, 2);
console.log('Graph adjacency list:');
graph.display();
